test/support/src/vfs_helpers.cc

```diff
@@ -415,5 +442,132 @@ std::string TemporaryDirectoryFixture::create_temporary_array(
   return array_uri;
 }
 
+VFSTest::VFSTest(
+    const std::vector<size_t>& test_tree, const std::string& prefix)
+    : test_tree_(test_tree)
+    , compute_(4)
+    , io_(4)
+    , vfs_(
+          &tiledb::test::g_helper_stats,
+          &io_,
+          &compute_,
+          tiledb::test::create_test_config())
+    , temp_dir_(tiledb::test::test_dir(prefix)) {
+  if (prefix == "file://") {
+#ifdef _WIN32
+    temp_dir_ =
+        tiledb::test::test_dir(prefix + tiledb::sm::Win::current_dir() + "/");
+#else
+    temp_dir_ =
+        tiledb::test::test_dir(prefix + tiledb::sm::Posix::current_dir() + "/");
+#endif
+  }
+
+  if (temp_dir_.is_file() || temp_dir_.is_memfs()) {
+    vfs_.create_dir(temp_dir_).ok();
+  } else if (vfs_.supports_uri_scheme(temp_dir_)) {
+    vfs_.create_bucket(temp_dir_).ok();
+  }
+}
+
+VFSTest::~VFSTest() {
+  if (vfs_.supports_uri_scheme(temp_dir_)) {
+    bool is_dir = false;
+    vfs_.is_dir(temp_dir_, &is_dir).ok();
+    if (is_dir) {
+      vfs_.remove_dir(temp_dir_).ok();
+    }
+  }
+}
+
+void VFSTest::create_objects(
+    const sm::URI& uri, size_t count, const std::string& prefix) {
+  for (size_t i = 1; i <= count; i++) {
+    auto object_uri = uri.join_path(prefix + std::to_string(i));
+    vfs_.touch(object_uri).ok();
+    std::string data(i * 10, 'a');
+    vfs_.open_file(object_uri, sm::VFSMode::VFS_WRITE).ok();
+    vfs_.write(object_uri, data.data(), data.size()).ok();
+    vfs_.close_file(object_uri).ok();
+    expected_results_.emplace_back(object_uri.to_string(), data.size());
+  }
+}
+
+void VFSTest::setup_test() {
+  for (size_t i = 1; i <= test_tree_.size(); i++) {
+    sm::URI path = temp_dir_.join_path("subdir_" + std::to_string(i));
+    // VFS::create_dir is a no-op for S3.
+    vfs_.create_dir(path).ok();
+    create_objects(path, test_tree_[i - 1], "test_file_");
+  }
+}
+
+void VFSTest::test_ls_recursive(
+    tiledb::sm::LsCallback cb, size_t expected_count) {
+  LsObjects ls_objects;
+  CHECK_NOTHROW(vfs_.ls_recursive(temp_dir_, cb, &ls_objects));
+
+  std::sort(expected_results_.begin(), expected_results_.end());
+  if (expected_count != 0) {
+    expected_results_.resize(expected_count);
+  }
+  CHECK(ls_objects.size() == expected_results_.size());
+  CHECK(ls_objects == expected_results_);
+}
+
+S3Test::S3Test(const std::vector<size_t>& test_tree)
+    : VFSTest(test_tree, "s3://") {
+}
+
+void S3Test::create_objects(
+    [[maybe_unused]] const sm::URI& uri,
+    [[maybe_unused]] size_t count,
+    [[maybe_unused]] const std::string& prefix) {
+#ifdef HAVE_S3
+  for (size_t i = 1; i <= count; i++) {
+    auto object_uri = uri.join_path(prefix + std::to_string(i));
+    vfs_.s3()->touch(object_uri).ok();
+    std::string data(i * 10, 'a');
+    vfs_.s3()->write(object_uri, data.data(), data.size()).ok();
+    vfs_.s3()->flush_object(object_uri).ok();
+    expected_results_.emplace_back(object_uri.to_string(), data.size());
+  }
+#endif
```

Comment:
If I move `setup_test()` into the constructor and throw here I wouldn't be able to construct `S3Test` without throwing for builds without S3 enabled, so the TEMPLATE_LIST_TEST_CASE would fail before we leave the constructor. I removed `create_objects` since it's only called by `setup_test` and used preproccessor checks to turn that into a stub if the backend is not enabled by the build. I left a separate comment on why I used preproccessor instead of constexpr.

Reply:
>I wouldn't be able to construct S3Test without throwing for builds without S3 enabled

Naively, yes, that would be the case. We've already solved exactly this problem with `struct S3_within_VFS`, from which `class VFS` gets its `s3_` member variable.. We'll need to use exactly this solution for `class VFSTest`. I think we should be able to use the same `struct `S3_within_VFS`.